architecture/2024-08-bicepdeployment-controller.md

```diff
@@ -0,0 +1,379 @@
+          "properties": {
+            "name": "container",
+            "properties": {
+              "application": "[parameters('application')]",
+              "container": {
+                "image": "[format('ghcr.io/radius-project/samples/demo:{0}', parameters('tag'))]",
+                "ports": {
+                  "web": {
+                    "containerPort": "[parameters('port')]"
+                  }
+                }
+              }
+            }
+          }
+        }
+      }
+    }
+  parameters: |
+    {
+      "application": {
+        "value": ""
+      },
+      "tag": {
+        "value": "latest"
+      },
+      "port": {
+        "value": 3000
+      }
+    }
+```
+
+## Design
+
+### High Level Design
+
+This design proposes the creation of two new Kubernetes custom resources and their corresponding reconcilers: `BicepDeployment` and `BicepResource`. The `BicepDeployment` controller will be responsible for deploying and reconciling resources defined in ARM JSON manifests via Radius, while the `BicepResource` controller will be responsible for reconciling individual resources. The `BicepDeployment` controller will have a control loop that will check the status of an in-progress operation, process deletion, and process creation or update. The `BicepResource` controller will have a similar control loop that will check the status of an in-progress operation or process deletion if necessary.
+
+### Architecture Diagram
+
+![Architecture Diagram](2024-08-bicepdeployment-controller/architecture.png)
+
+### Detailed Design
+
+#### BicepDeployment Custom Resource Definition
+
+The `BicepDeployment` CRD will be a new Kubernetes CRD that will be used to contain the ARM JSON manifest and parameters. The CRD will have the following fields:
+
+```go
+// BicepDeploymentSpec defines the desired state of a BicepDeployment
+type BicepDeploymentSpec struct {
+  // Template is the ARM JSON manifest that defines the resources to deploy.
+  Template string `json:"template"`
+
+  // Parameters is the ARM JSON parameters for the template.
+  Parameters string `json:"parameters"`
+}
+
+// BicepDeploymentStatus defines the observed state of the Bicep Deployment.
+type BicepDeploymentStatus struct {
+  // ObservedGeneration is the most recent generation observed for this BicepDeployment.
+  ObservedGeneration int64 `json:"observedGeneration,omitempty"`
+
+	// Scope is the resource ID of the scope.
+	Scope string `json:"scope,omitempty"`
+
+	// Resource is the resource ID of the deployment.
+	Resource string `json:"resource,omitempty"`
+
+	// Operation tracks the status of an in-progress provisioning operation.
+	Operation *ResourceOperation `json:"operation,omitempty"`
+
+	// Phrase indicates the current status of the Bicep Deployment.
+	Phrase BicepDeploymentPhrase `json:"phrase,omitempty"`
+
+  // Message is a human-readable description of the status of the Bicep Deployment.
+  Message string `json:"message,omitempty"`
+}
```

Comment:
Where are errors reported?

Reply:
update what gets described on kubectl get and kubectl describe

Reply:
Do we store UCP-structured error messages or parse it